File: multi-region/03-pricing-optimized/pricing_optimized.ts
/**
 * Pricing-optimized multi-region deployment for HarchOS (TypeScript).
 *
 * Deploys workloads across regions optimized for cost while respecting
 * carbon intensity thresholds and sovereignty requirements.
 *
 * Usage:
 *   npx ts-node pricing_optimized.ts --gpu-type H100 --gpu-count 8 --hours 168
 *   npx ts-node pricing_optimized.ts --carbon-max 100
 */

import HarchOS from "@harchos/sdk";
import type { Region, PricingPlan, EstimateCostParams } from "harchos";

// ---------------------------------------------------------------------------
// Simulated data
// ---------------------------------------------------------------------------

interface HubInfo {
  name: string;
  region: string;
  gpus: number;
  renewablePct: number;
  carbonIntensity: number;
  gpuType: string;
  pricePerGpuHour: number;
  tier: string;
}

const SIMULATED_HUBS: HubInfo[] = [
  { name: "Harch Ouarzazate", region: "morocco", gpus: 800, renewablePct: 97.2, carbonIntensity: 18, gpuType: "H100", pricePerGpuHour: 2.10, tier: "enterprise" },
  { name: "Harch Dakhla", region: "morocco", gpus: 400, renewablePct: 94.8, carbonIntensity: 32, gpuType: "L40S", pricePerGpuHour: 1.40, tier: "enterprise" },
  { name: "Harch Benguerir", region: "morocco", gpus: 350, renewablePct: 88.5, carbonIntensity: 55, gpuType: "H100", pricePerGpuHour: 2.35, tier: "performance" },
  { name: "Harch Tanger", region: "morocco", gpus: 200, renewablePct: 82.1, carbonIntensity: 95, gpuType: "A100", pricePerGpuHour: 1.80, tier: "enterprise" },
  { name: "Harch Casablanca", region: "morocco", gpus: 48, renewablePct: 45.0, carbonIntensity: 210, gpuType: "A100", pricePerGpuHour: 1.95, tier: "community" },
];

const SOVEREIGNTY_RULES: Record<string, { frameworks: string[]; dataResidency: boolean }> = {
  morocco: { frameworks: ["GDPR", "CNDP"], dataResidency: true },
  nigeria: { frameworks: ["NDPR"], dataResidency: true },
  kenya: { frameworks: ["PDPA"], dataResidency: true },
};

// ---------------------------------------------------------------------------
// Logic
// ---------------------------------------------------------------------------

interface DeploymentRecommendation {
  recommendation: "deploy" | "relax_constraints" | "no_suitable_hub";
  hub?: string;
  region?: string;
  gpuType?: string;
  gpuCount?: number;
  pricePerGpuHour?: number;
  carbonIntensity?: number;
  renewablePct?: number;
  estimatedCost?: number;
  hours?: number;
  tier?: string;
  alternatives?: Array<{
    hub: string;
    pricePerGpuHour: number;
    carbonIntensity: number;
    estimatedCost: number;
  }>;
  sovereignty?: { frameworks: string[]; dataResidency: boolean };
  reason?: string;
  suggestion?: string;
}

function getEligibleHubs(
  gpuType: string,
  gpuCount: number,
  carbonMax: number,
  tier?: string,
  sovereignOnly?: boolean,
): HubInfo[] {
  let eligible = SIMULATED_HUBS.filter((h) => {
    if (h.gpuType !== gpuType) return false;
    if (h.gpus < gpuCount) return false;
    if (h.carbonIntensity > carbonMax) return false;
    if (tier && h.tier !== tier) return false;
    if (sovereignOnly) {
      const rules = SOVEREIGNTY_RULES[h.region];
      if (!rules?.dataResidency) return false;
    }
    return true;
  });

  eligible.sort((a, b) => a.pricePerGpuHour - b.pricePerGpuHour);
  return eligible;
}

function recommend(
  gpuType: string,
  gpuCount: number,
  hours: number,
  carbonMax: number,
  tier?: string,
  sovereignOnly?: boolean,
): DeploymentRecommendation {
  const eligible = getEligibleHubs(gpuType, gpuCount, carbonMax, tier, sovereignOnly);

  if (!eligible.length) {
    const allMatching = SIMULATED_HUBS.filter(
      (h) => h.gpuType === gpuType && h.gpus >= gpuCount
    );
    if (allMatching.length) {
      const cheapest = allMatching.reduce((a, b) =>
        a.pricePerGpuHour < b.pricePerGpuHour ? a : b
      );
      return {
        recommendation: "relax_constraints",
        reason: `No hubs within carbon limit (${carbonMax} gCO2/kWh)`,
        suggestion: `Increase --carbon-max to at least ${cheapest.carbonIntensity}`,
      };
    }
    return {
      recommendation: "no_suitable_hub",
      reason: `No hubs with ${gpuCount}x ${gpuType} GPUs available`,
    };
  }

  const best = eligible[0];
  const estimatedCost = best.pricePerGpuHour * gpuCount * hours;

  return {
    recommendation: "deploy",
    hub: best.name,
    region: best.region,
    gpuType: best.gpuType,
    gpuCount,
    pricePerGpuHour: best.pricePerGpuHour,
    carbonIntensity: best.carbonIntensity,
    renewablePct: best.renewablePct,
    estimatedCost: Math.round(estimatedCost * 100) / 100,
    hours,
    tier: best.tier,
    alternatives: eligible.slice(1).map((h) => ({
      hub: h.name,
      pricePerGpuHour: h.pricePerGpuHour,
      carbonIntensity: h.carbonIntensity,
      estimatedCost: Math.round(h.pricePerGpuHour * gpuCount * hours * 100) / 100,
    })),
    sovereignty: SOVEREIGNTY_RULES[best.region],
  };
}

function printRecommendation(rec: DeploymentRecommendation): void {
  if (rec.recommendation === "no_suitable_hub") {
    console.log(`\n  ❌ ${rec.reason}`);
    return;
  }

  if (rec.recommendation === "relax_constraints") {
    console.log(`\n  ⚠️  ${rec.reason}`);
    console.log(`     Suggestion: ${rec.suggestion}`);
    return;
  }

  console.log("\n  🎯 Deployment Recommendation");
  console.log("  " + "─".repeat(60));
  console.log(`    Hub:               ${rec.hub}`);
  console.log(`    Region:            ${rec.region}`);
  console.log(`    GPU:               ${rec.gpuCount}x ${rec.gpuType}`);
  console.log(`    Tier:              ${rec.tier}`);
  console.log(`    Duration:          ${rec.hours} hours`);
  console.log(`    Price/GPU/hr:      $${rec.pricePerGpuHour?.toFixed(2)}`);
  console.log(`    Carbon Intensity:  ${rec.carbonIntensity} gCO2/kWh`);
  console.log(`    Renewable:         ${rec.renewablePct?.toFixed(1)}%`);
  console.log(`    Estimated Cost:    $${rec.estimatedCost?.toFixed(2)} USD`);

  if (rec.sovereignty) {
    console.log("\n  🛡️  Sovereignty");
    console.log(`    Data Residency:    ${rec.sovereignty.dataResidency ? "Required ✅" : "Optional"}`);
    if (rec.sovereignty.frameworks.length) {
      console.log(`    Compliance:        ${rec.sovereignty.frameworks.join(", ")}`);
    }
  }

  if (rec.alternatives?.length) {
    console.log("\n  🔄 Alternatives (sorted by price)");
    console.log("  " + "─".repeat(60));
    for (const alt of rec.alternatives) {
      const icon = alt.carbonIntensity <= 100 ? "🟢" : alt.carbonIntensity <= 200 ? "🟡" : "🔴";
      console.log(
        `    ${icon} ${alt.hub.padEnd(25)} | ` +
        `$${alt.pricePerGpuHour.toFixed(2)}/gpu/hr | ` +
        `${alt.carbonIntensity} gCO2/kWh | ` +
        `$${alt.estimatedCost.toFixed(2)}`
      );
    }
  }
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  const gpuType = getArg(args, "--gpu-type", "H100");
  const gpuCount = parseInt(getArg(args, "--gpu-count", "4"), 10);
  const hours = parseFloat(getArg(args, "--hours", "24"));
  const carbonMax = parseFloat(getArg(args, "--carbon-max", "200"));
  const tier = getArg(args, "--tier", "") || undefined;
  const sovereignOnly = args.includes("--sovereign-only");

  console.log("=".repeat(60));
  console.log("HarchOS Pricing-Optimized Deployment (TypeScript)");
  console.log("=".repeat(60));

  // Try SDK for regions
  const apiKey = process.env.HARCHOS_API_KEY || "";
  if (apiKey) {
    try {
      const client = new HarchOS({ apiKey });
      const regions = await client.regions.list({ available: true });
      console.log(`\n  📍 Available Regions (from API): ${regions.length}`);
      for (const r of regions) {
        console.log(`    • ${r.name} (${r.code}): ${r.hubCount} hubs, ${r.totalGpus} GPUs`);
      }
    } catch (err) {
      console.log(`  [WARN] SDK call failed: ${err}`);
    }
  }

  console.log(`\n  🔍 Finding best hub for ${gpuCount}x ${gpuType} (${hours}h, carbon ≤ ${carbonMax} gCO2/kWh)...`);

  const rec = recommend(gpuType, gpuCount, hours, carbonMax, tier, sovereignOnly);
  printRecommendation(rec);
  console.log();
}

function getArg(args: string[], flag: string, defaultValue: string): string {
  const idx = args.indexOf(flag);
  if (idx !== -1 && idx + 1 < args.length) return args[idx + 1];
  return defaultValue;
}

main().catch((err) => {
  console.error("Fatal error:", err);
  process.exit(1);
});

File: multi-region/03-pricing-optimized/pricing_optimized.py
#!/usr/bin/env python3
"""Pricing-optimized multi-region deployment for HarchOS.

Deploys workloads across HarchOS regions optimized for cost while
respecting carbon intensity thresholds and sovereignty requirements.
Combines the pricing and regions APIs to find the best-value hub
for each workload.

Features:
- Cross-region cost comparison before deployment
- Carbon-intensity constraints on hub selection
- Sovereignty enforcement (data residency compliance)
- Tier-aware pricing optimization
- Deployment recommendation engine

Usage:
    python pricing_optimized.py --gpu-type H100 --gpu-count 8 --hours 168
    python pricing_optimized.py --carbon-max 100 --tier enterprise
    python pricing_optimized.py --sovereign-only
    HARCHOS_API_KEY=hsk_... python pricing_optimized.py
"""

import argparse
import json
import os
from datetime import datetime, timezone
from typing import Dict, List, Optional

try:
    from harchos import HarchOS
    HARCHOS_AVAILABLE = True
except ImportError:
    HARCHOS_AVAILABLE = False
    print("[WARN] harchos SDK not installed — using simulated data")


# ---------------------------------------------------------------------------
# Simulated data
# ---------------------------------------------------------------------------

SIMULATED_REGIONS = [
    {
        "name": "Morocco", "code": "morocco", "country": "MA",
        "available": True, "hub_count": 5, "total_gpus": 1798,
        "avg_renewable_percentage": 81.52, "avg_carbon_intensity": 47.0,
        "latency_ms_from_casablanca": 0,
        "compliance_frameworks": ["GDPR", "CNDP", "ISO27001"],
    },
    {
        "name": "Nigeria", "code": "nigeria", "country": "NG",
        "available": False, "hub_count": 0, "total_gpus": 0,
        "avg_renewable_percentage": 0, "avg_carbon_intensity": 0,
        "latency_ms_from_casablanca": 85,
        "compliance_frameworks": ["NDPR", "GDPR"],
    },
    {
        "name": "Kenya", "code": "kenya", "country": "KE",
        "available": False, "hub_count": 0, "total_gpus": 0,
        "avg_renewable_percentage": 0, "avg_carbon_intensity": 0,
        "latency_ms_from_casablanca": 120,
        "compliance_frameworks": ["PDPA", "GDPR"],
    },
]

SIMULATED_HUBS = [
    {"name": "Harch Ouarzazate", "region": "morocco", "gpus": 800,
     "renewable_pct": 97.2, "carbon_intensity": 18,
     "gpu_type": "H100", "price_per_gpu_hour": 2.10, "tier": "enterprise"},
    {"name": "Harch Dakhla", "region": "morocco", "gpus": 400,
     "renewable_pct": 94.8, "carbon_intensity": 32,
     "gpu_type": "L40S", "price_per_gpu_hour": 1.40, "tier": "enterprise"},
    {"name": "Harch Benguerir", "region": "morocco", "gpus": 350,
     "renewable_pct": 88.5, "carbon_intensity": 55,
     "gpu_type": "H100", "price_per_gpu_hour": 2.35, "tier": "performance"},
    {"name": "Harch Tanger", "region": "morocco", "gpus": 200,
     "renewable_pct": 82.1, "carbon_intensity": 95,
     "gpu_type": "A100", "price_per_gpu_hour": 1.80, "tier": "enterprise"},
    {"name": "Harch Casablanca", "region": "morocco", "gpus": 48,
     "renewable_pct": 45.0, "carbon_intensity": 210,
     "gpu_type": "A100", "price_per_gpu_hour": 1.95, "tier": "community"},
]


# ---------------------------------------------------------------------------
# Pricing-Optimized Deployer
# ---------------------------------------------------------------------------

class PricingOptimizedDeployer:
    """Deploys workloads to the most cost-effective hub across regions."""

    # Required sovereignty frameworks per region
    SOVEREIGNTY_RULES = {
        "morocco": {"frameworks": ["GDPR", "CNDP"], "data_residency": True},
        "nigeria": {"frameworks": ["NDPR"], "data_residency": True},
        "kenya": {"frameworks": ["PDPA"], "data_residency": True},
    }

    def __init__(self, carbon_max: float = 200.0, tier: Optional[str] = None,
                 sovereign_only: bool = False):
        self.carbon_max = carbon_max
        self.tier = tier
        self.sovereign_only = sovereign_only

        if HARCHOS_AVAILABLE:
            self.client = HarchOS(api_key=os.environ.get("HARCHOS_API_KEY", ""))
        else:
            self.client = None

    def get_regions(self) -> List[dict]:
        """Fetch available regions."""
        if self.client:
            try:
                regions = self.client.regions.list(available=True)
                return [
                    {
                        "name": r.name, "code": r.code, "country": r.country,
                        "available": r.available, "hub_count": r.hub_count,
                        "total_gpus": r.total_gpus,
                        "avg_renewable_percentage": r.avg_renewable_percentage,
                        "avg_carbon_intensity": r.avg_carbon_intensity,
                        "latency_ms_from_casablanca": r.latency_ms_from_casablanca or 0,
                        "compliance_frameworks": r.compliance_frameworks,
                    }
                    for r in regions
                ]
            except Exception:
                pass

        return [r for r in SIMULATED_REGIONS if r["available"]]

    def get_eligible_hubs(self, gpu_type: str, gpu_count: int) -> List[dict]:
        """Find hubs that match GPU type, count, carbon, and tier constraints."""
        eligible = []

        for hub in SIMULATED_HUBS:
            # GPU type check
            if hub["gpu_type"] != gpu_type:
                continue

            # GPU count check
            if hub["gpus"] < gpu_count:
                continue

            # Carbon check
            if hub["carbon_intensity"] > self.carbon_max:
                continue

            # Tier check
            if self.tier and hub["tier"] != self.tier:
                continue

            # Sovereignty check
            if self.sovereign_only:
                region_code = hub["region"]
                rules = self.SOVEREIGNTY_RULES.get(region_code, {})
                if not rules.get("data_residency", False):
                    continue

            eligible.append(hub)

        # Sort by price (ascending)
        eligible.sort(key=lambda h: h["price_per_gpu_hour"])
        return eligible

    def recommend_deployment(
        self,
        gpu_type: str,
        gpu_count: int,
        hours: float,
    ) -> dict:
        """Recommend the best hub for deployment based on price + carbon."""
        regions = self.get_regions()
        eligible_hubs = self.get_eligible_hubs(gpu_type, gpu_count)

        if not eligible_hubs:
            # Try relaxing constraints
            all_matching = [
                h for h in SIMULATED_HUBS
                if h["gpu_type"] == gpu_type and h["gpus"] >= gpu_count
            ]
            if all_matching:
                cheapest = min(all_matching, key=lambda h: h["price_per_gpu_hour"])
                return {
                    "recommendation": "relax_constraints",
                    "reason": f"No hubs within carbon limit ({self.carbon_max} gCO2/kWh)",
                    "cheapest_available": cheapest["name"],
                    "cheapest_carbon": cheapest["carbon_intensity"],
                    "suggestion": f"Increase --carbon-max to at least {cheapest['carbon_intensity']}",
                }
            return {
                "recommendation": "no_suitable_hub",
                "reason": f"No hubs with {gpu_count}x {gpu_type} GPUs available",
            }

        # Best option: cheapest eligible hub
        best = eligible_hubs[0]
        estimated_cost = best["price_per_gpu_hour"] * gpu_count * hours

        return {
            "recommendation": "deploy",
            "hub": best["name"],
            "region": best["region"],
            "gpu_type": best["gpu_type"],
            "gpu_count": gpu_count,
            "price_per_gpu_hour": best["price_per_gpu_hour"],
            "carbon_intensity": best["carbon_intensity"],
            "renewable_pct": best["renewable_pct"],
            "estimated_cost": round(estimated_cost, 2),
            "currency": "USD",
            "hours": hours,
            "tier": best["tier"],
            "alternatives": [
                {
                    "hub": h["name"],
                    "price_per_gpu_hour": h["price_per_gpu_hour"],
                    "carbon_intensity": h["carbon_intensity"],
                    "estimated_cost": round(h["price_per_gpu_hour"] * gpu_count * hours, 2),
                }
                for h in eligible_hubs[1:]
            ],
            "sovereignty": self.SOVEREIGNTY_RULES.get(best["region"], {}),
        }

    def print_recommendation(self, rec: dict):
        """Pretty-print the deployment recommendation."""
        if rec["recommendation"] == "no_suitable_hub":
            print(f"\n  ❌ {rec['reason']}")
            return

        if rec["recommendation"] == "relax_constraints":
            print(f"\n  ⚠️  {rec['reason']}")
            print(f"     Cheapest available: {rec['cheapest_available']} "
                  f"({rec['cheapest_carbon']} gCO2/kWh)")
            print(f"     Suggestion: {rec['suggestion']}")
            return

        print(f"\n  🎯 Deployment Recommendation")
        print(f"  {'─' * 60}")
        print(f"    Hub:               {rec['hub']}")
        print(f"    Region:            {rec['region']}")
        print(f"    GPU:               {rec['gpu_count']}x {rec['gpu_type']}")
        print(f"    Tier:              {rec['tier']}")
        print(f"    Duration:          {rec['hours']} hours")
        print(f"    Price/GPU/hr:      ${rec['price_per_gpu_hour']:.2f}")
        print(f"    Carbon Intensity:  {rec['carbon_intensity']} gCO2/kWh")
        print(f"    Renewable:         {rec['renewable_pct']:.1f}%")
        print(f"    Estimated Cost:    ${rec['estimated_cost']:.2f} {rec['currency']}")

        # Sovereignty info
        sovereignty = rec.get("sovereignty", {})
        if sovereignty:
            print(f"\n  🛡️  Sovereignty")
            print(f"    Data Residency:    {'Required ✅' if sovereignty.get('data_residency') else 'Optional'}")
            frameworks = sovereignty.get("frameworks", [])
            if frameworks:
                print(f"    Compliance:        {', '.join(frameworks)}")

        # Alternatives
        alternatives = rec.get("alternatives", [])
        if alternatives:
            print(f"\n  🔄 Alternatives (sorted by price)")
            print(f"  {'─' * 60}")
            for alt in alternatives:
                carbon_icon = "🟢" if alt["carbon_intensity"] <= 100 else (
                    "🟡" if alt["carbon_intensity"] <= 200 else "🔴")
                print(f"    {carbon_icon} {alt['hub']:25s} | "
                      f"${alt['price_per_gpu_hour']:.2f}/gpu/hr | "
                      f"{alt['carbon_intensity']} gCO2/kWh | "
                      f"${alt['estimated_cost']:.2f}")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main():
    parser = argparse.ArgumentParser(description="Pricing-Optimized Multi-Region Deployment")
    parser.add_argument("--gpu-type", type=str, default="H100",
                        help="GPU type (H100, A100, L40S)")
    parser.add_argument("--gpu-count", type=int, default=4,
                        help="Number of GPUs needed")
    parser.add_argument("--hours", type=float, default=24.0,
                        help="Estimated usage hours")
    parser.add_argument("--carbon-max", type=float, default=200.0,
                        help="Maximum carbon intensity in gCO2/kWh")
    parser.add_argument("--tier", type=str, default=None,
                        choices=["community", "enterprise", "sovereign"],
                        help="Required pricing tier")
    parser.add_argument("--sovereign-only", action="store_true",
                        help="Only consider sovereign-compliant hubs")
    args = parser.parse_args()

    deployer = PricingOptimizedDeployer(
        carbon_max=args.carbon_max,
        tier=args.tier,
        sovereign_only=args.sovereign_only,
    )

    print("=" * 60)
    print("HarchOS Pricing-Optimized Multi-Region Deployment")
    print("=" * 60)

    # Show available regions
    regions = deployer.get_regions()
    print(f"\n  📍 Available Regions: {len(regions)}")
    for r in regions:
        print(f"    • {r['name']} ({r['code']}): {r['hub_count']} hubs, "
              f"{r['total_gpus']} GPUs, {r['avg_renewable_percentage']:.1f}% renewable")

    # Get recommendation
    print(f"\n  🔍 Finding best hub for {args.gpu_count}x {args.gpu_type} "
          f"({args.hours}h, carbon ≤ {args.carbon_max} gCO2/kWh)...")

    rec = deployer.recommend_deployment(
        gpu_type=args.gpu_type,
        gpu_count=args.gpu_count,
        hours=args.hours,
    )
    deployer.print_recommendation(rec)
    print()


if __name__ == "__main__":
    main()
